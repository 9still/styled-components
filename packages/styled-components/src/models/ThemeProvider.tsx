import React, { useContext, useMemo } from 'react';
import styledError from '../utils/error';
import isFunction from '../utils/isFunction';

export const MQ_THEME_KEY = '_sc_media';

export type Theme = {
  [MQ_THEME_KEY]?: {
    [key: string]: any;
  };
  [key: string]: any;
};

type ThemeFn = (outerTheme?: Theme) => Theme;
type ThemeArgument = Theme | ThemeFn;

type Props = {
  children?: React.ReactChild;
  theme: ThemeArgument;
};

export const ThemeContext = React.createContext<Theme | undefined>(undefined);

export const ThemeConsumer = ThemeContext.Consumer;

function mergeTheme(theme: ThemeArgument, outerTheme?: Theme): Theme {
  if (!theme) {
    throw styledError(14);
  }

  if (isFunction(theme)) {
    const themeFn = theme as ThemeFn;
    const mergedTheme = themeFn(outerTheme);

    if (
      process.env.NODE_ENV !== 'production' &&
      (mergedTheme === null || Array.isArray(mergedTheme) || typeof mergedTheme !== 'object')
    ) {
      throw styledError(7);
    }

    return mergedTheme;
  }

  if (Array.isArray(theme) || typeof theme !== 'object') {
    throw styledError(8);
  }

  return outerTheme ? { ...outerTheme, ...theme } : theme;
}

function synthesizeCSSVariables(obj: { [key: string]: any }, parentKey = ''): string {
  let vars = '';
  let needsToEmitMediaQueries = false;

  if (!parentKey) {
    vars += ':root{';
  }

  for (const key in obj) {
    if (obj.hasOwnProperty(key)) {
      if (key === MQ_THEME_KEY) {
        needsToEmitMediaQueries = true;
      } else if (typeof obj[key] !== 'object') {
        vars += `${normalizeKeyName(parentKey ? `${parentKey}.${key}` : key)}: ${obj[key]};`;
      } else {
        vars += synthesizeCSSVariables(obj[key], key);
      }
    }
  }

  if (!parentKey) vars += '}';

  if (needsToEmitMediaQueries) {
    for (const query in obj[MQ_THEME_KEY]) {
      if (obj[MQ_THEME_KEY].hasOwnProperty(query)) {
        vars += `@media (${query}){${synthesizeCSSVariables(obj[MQ_THEME_KEY][query])}}\n`;
      }
    }
  }

  return vars;
}

function normalizeKeyName(key: string): string {
  return `--sc-${key
    .toLowerCase()
    .split('.')
    .map(x => x.replace(/[^A-Z0-9]/gi, '-').replace(/-{2,}/g, '-'))
    .join('-')}`;
}

// thank you https://www.reddit.com/r/typescript/comments/iywewf/comment/g6fh1m5/?utm_source=share&utm_medium=web2x&context=3
type PropsPath<T extends Theme> = {
  [P in keyof T]: T[P] extends Theme
    ? `${string & P}` | `${string & P}.${PropsPath<T[P]>}`
    : `${string & P}`;
}[T extends any[] ? number & keyof T : keyof T];

/**
 * Given a theme key path (dot-delimited like JS), consume the appropriate emitted CSS variable
 * taking into account any provided media query adjustments.
 *
 * e.g.
 *
 * ```
 * styled.div`
 *   color: ${cssvar('color')};
 * `
 * ```
 */
export function cssvar<T = Theme>(key: PropsPath<T>): string {
  // @ts-expect-error it's fine
  return `var(${normalizeKeyName(String(key))})`;
}

/**
 * Provide a theme to an entire react component tree via context
 */
export default function ThemeProvider(props: Props): JSX.Element | null {
  const outerTheme = useContext(ThemeContext);
  const themeContext = useMemo(
    () => mergeTheme(props.theme, outerTheme),
    [props.theme, outerTheme]
  );

  /**
   * Synthesize CSS variables from theme.
   */
  const cssVars = useMemo(() => {
    if (themeContext) {
      return synthesizeCSSVariables(themeContext);
    }
  }, [themeContext]);

  if (!props.children) {
    return null;
  }

  return (
    <ThemeContext.Provider value={themeContext}>
      {cssVars && <style dangerouslySetInnerHTML={{ __html: cssVars }} />}
      {props.children}
    </ThemeContext.Provider>
  );
}
