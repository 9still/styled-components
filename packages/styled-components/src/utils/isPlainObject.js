// @flow

export default (x: any): boolean =>
  x !== null &&
  typeof x === 'object' &&
  (x.toString ? x.toString() : Object.prototype.toString.call(x)) === '[object Object]' &&
  /* check for reasonable markers that the object isn't an element for react & preact/compat */
  !(x.$$typeof && (x._owner || (x.__v && x.__o)));
