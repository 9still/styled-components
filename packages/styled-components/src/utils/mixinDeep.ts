import { ExtensibleObject } from '../types';
import isPlainObject from './isPlainObject'

function isRecursible(obj: any) {
  return isPlainObject(obj) || Array.isArray(obj);
}

function mixinRecursively(target: any, source: any) {
  const isRecursive = isRecursible(target);
  for (const [key, sourceVal] of Object.entries(source)) {
    const targetVal = target[key];
    if (isRecursive && isRecursible(targetVal) && isRecursible(sourceVal)) {
      target[key] = mixinRecursively(targetVal, sourceVal);
    } else {
      target[key] = sourceVal;
    }
  };

  return target;
}

/**
 * Arrays & POJOs merged recursively, other objects and value types are overridden
 * Source objects applied left to right.  Mutates & returns target.  Similar to lodash merge.
 */
export default function mixinDeep(target: ExtensibleObject = {}, ...sources: any[]) {
  for (const source of sources) {
    if (isRecursible(source)) {
      mixinRecursively(target, source);
    }
  };
  return target;
}
