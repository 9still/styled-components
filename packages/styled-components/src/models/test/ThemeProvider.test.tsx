/* eslint-disable react/no-multi-comp */
import React from 'react';
import TestRenderer from 'react-test-renderer';
import withTheme from '../../hoc/withTheme';
import { expectCSSMatches, resetStyled } from '../../test/utils';
import ThemeProvider, { cssvar, MQ_THEME_KEY } from '../ThemeProvider';

let styled: ReturnType<typeof resetStyled>;

describe('ThemeProvider', () => {
  beforeEach(() => {
    styled = resetStyled();
  });

  it('should not throw an error when no children are passed', () => {
    TestRenderer.create(<ThemeProvider theme={{}} />);
  });

  it("should accept a theme prop that's a plain object", () => {
    TestRenderer.create(<ThemeProvider theme={{ main: 'black' }} />);
  });

  it('should render its child', () => {
    const child = <p>Child!</p>;
    const wrapper = TestRenderer.create(
      <ThemeProvider theme={{ main: 'black' }}>{child}</ThemeProvider>
    );

    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should merge its theme with an outer theme', () => {
    const outerTheme = { main: 'black' };
    const innerTheme = { secondary: 'black' };

    const MyDiv = styled.div``;
    const MyDivWithTheme = withTheme(MyDiv);

    const wrapper = TestRenderer.create(
      <ThemeProvider theme={outerTheme}>
        <ThemeProvider theme={innerTheme}>
          <MyDivWithTheme />
        </ThemeProvider>
      </ThemeProvider>
    );

    expect(wrapper.root.findByType(MyDiv).props.theme).toEqual({
      ...outerTheme,
      ...innerTheme,
    });
  });

  it('should merge its theme with multiple outer themes', () => {
    const outerestTheme = { main: 'black' };
    const outerTheme = { main: 'blue' };
    const innerTheme = { secondary: 'black' };

    const MyDiv = styled.div``;
    const MyDivWithTheme = withTheme(MyDiv);

    const wrapper = TestRenderer.create(
      <ThemeProvider theme={outerestTheme}>
        <ThemeProvider theme={outerTheme}>
          <ThemeProvider theme={innerTheme}>
            <MyDivWithTheme />
          </ThemeProvider>
        </ThemeProvider>
      </ThemeProvider>
    );

    expect(wrapper.root.findByType(MyDiv).props.theme).toEqual({
      ...outerestTheme,
      ...outerTheme,
      ...innerTheme,
    });
  });

  it('should be able to render two independent themes', () => {
    const themes = {
      one: { main: 'black', secondary: 'red' },
      two: { main: 'blue', other: 'green' },
    };

    const MyDivOne = withTheme(styled.div``);
    const MyDivWithThemeOne = withTheme(MyDivOne);
    const MyDivTwo = withTheme(styled.div``);
    const MyDivWithThemeTwo = withTheme(MyDivTwo);

    const wrapper = TestRenderer.create(
      <div>
        <ThemeProvider theme={themes.one}>
          <MyDivWithThemeOne />
        </ThemeProvider>
        <ThemeProvider theme={themes.two}>
          <MyDivWithThemeTwo />
        </ThemeProvider>
      </div>
    );

    expect(wrapper.root.findByType(MyDivOne).props.theme).toEqual(themes.one);
    expect(wrapper.root.findByType(MyDivTwo).props.theme).toEqual(themes.two);
  });

  it('ThemeProvider propagates theme updates through nested ThemeProviders', () => {
    const theme = { themed: true };
    const augment = (outerTheme: typeof theme) =>
      Object.assign({}, outerTheme, { augmented: true });
    const update = { updated: true };
    const expected = { themed: true, augmented: true, updated: true };

    const MyDiv = styled.div``;
    const MyDivWithTheme = withTheme(MyDiv);

    const getJSX = (givenTheme = theme) => (
      <ThemeProvider theme={givenTheme}>
        <ThemeProvider theme={augment}>
          <MyDivWithTheme />
        </ThemeProvider>
      </ThemeProvider>
    );

    const wrapper = TestRenderer.create(getJSX());

    wrapper.update(getJSX(Object.assign({}, theme, update)));

    expect(wrapper.root.findByType(MyDiv).props.theme).toEqual(expected);
  });

  it('ThemeProvider emits the theme as CSS variables as well', () => {
    const wrapper = TestRenderer.create(
      <ThemeProvider
        theme={{
          [MQ_THEME_KEY]: { 'prefers-color-scheme: dark': { color: 'white' } },
          color: 'red',
          style: { background: 'blue' },
        }}
      >
        <div />
      </ThemeProvider>
    );

    expect(wrapper.toJSON()).toMatchInlineSnapshot(`
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ":root{--sc-color: red;--sc-style-background: blue;}@media (prefers-color-scheme: dark){:root{--sc-color: white;}}
",
      }
    }
  />,
  <div />,
]
`);
  });

  it('css variable synthesis emits media query variants properly', () => {
    const wrapper = TestRenderer.create(
      <ThemeProvider theme={{ color: 'red', style: { background: 'blue' } }}>
        <div />
      </ThemeProvider>
    );

    expect(wrapper.toJSON()).toMatchInlineSnapshot(`
  Array [
    <style
      dangerouslySetInnerHTML={
        Object {
          "__html": ":root{--sc-color: red;--sc-style-background: blue;}",
        }
      }
    />,
    <div />,
  ]
  `);
  });

  it('cssvar helper accesses theme properties correctly', () => {
    const theme = { color: 'red', style: { background: 'blue' } };

    const MyDiv = styled.div`
      color: ${cssvar<typeof theme>('color')};
    `;

    const wrapper = TestRenderer.create(
      <ThemeProvider theme={theme}>
        <MyDiv />
      </ThemeProvider>
    );

    expectCSSMatches('.b { color:var(--sc-color); }');
    expect(wrapper.toJSON()).toMatchInlineSnapshot(`
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ":root{--sc-color: red;--sc-style-background: blue;}",
      }
    }
  />,
  <div
    className="sc-a b"
  />,
]
`);
  });
});
