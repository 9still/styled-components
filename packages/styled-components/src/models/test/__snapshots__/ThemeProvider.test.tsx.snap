// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider should render its child 1`] = `
Array [
  <style
    dangerouslySetInnerHTML={
      Object {
        "__html": ":root{--sc-main: black;}",
      }
    }
  />,
  <p>
    Child!
  </p>,
]
`;
